# Roberto Cruz- Exercise 1

In [3]:
import numpy as np
import pandas as pd

## Problem 1

### 1.1

In [56]:
heartFailure_df = pd.read_csv('heart_failure_dataset.csv')

In [3]:
heartShape = heartFailure_df.shape
heartAxes = heartFailure_df.axes
heartDtype = heartFailure_df.dtypes

print("The shape of the dataset: ", heartShape)
print("\nThe axes of the dataset: \n", heartAxes)
print("\nThe dtype of the dataset: \n", heartDtype)

The shape of the dataset:  (299, 13)

The axes of the dataset: 
 [RangeIndex(start=0, stop=299, step=1), Index(['age', 'anaemia', 'creatinine_phosphokinase', 'diabetes',
       'ejection_fraction', 'high_blood_pressure', 'platelets',
       'serum_creatinine', 'serum_sodium', 'sex', 'smoking', 'time',
       'DEATH_EVENT'],
      dtype='object')]

The dtype of the dataset: 
 age                         float64
anaemia                       int64
creatinine_phosphokinase      int64
diabetes                      int64
ejection_fraction             int64
high_blood_pressure           int64
platelets                   float64
serum_creatinine            float64
serum_sodium                  int64
sex                           int64
smoking                       int64
time                          int64
DEATH_EVENT                   int64
dtype: object


In [4]:
print("First 10 rows of the dataset: \n", heartFailure_df.head(10))

First 10 rows of the dataset: 
     age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0  75.0        0                       582         0                 20   
1  55.0        0                      7861         0                 38   
2  65.0        0                       146         0                 20   
3  50.0        1                       111         0                 20   
4  65.0        1                       160         1                 20   
5  90.0        1                        47         0                 40   
6  75.0        1                       246         0                 15   
7  60.0        1                       315         1                 60   
8  65.0        0                       157         0                 65   
9  80.0        1                       123         0                 35   

   high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                    1  265000.00               1.9           130   

Each row is a data sample

From the *Death Event* column, this should be a supervised learning problem, more specifically a binary classification problem. I say this because while skimming through the datase, in that column there are 0s and 1s, which could potentially mean dead or alive- classifying between these two states from the other columns. 

There are 12 independent variables: 
<ol>
    <li>age</li>
    <li>anaemia</li>
    <li>creatine_phosphokinase</li>
    <li>diabetes</li>
    <li>ejection_fraction</li>
    <li>high_blood_pressure</li>
    <li>platelets</li>
    <li>serum_creatine</li>
    <li>serum_sodium</li>
    <li>sex</li>
    <li>smoking</li>
    <li>time</li>
</ol>

This dataset has 1 dependent variables: *Death Event*

There are 5 variables which are categorical:
<ul>
    <li>anaemia</li>
    <li>diabetes</li>
    <li>high_blood_pressure</li>
    <li>sex</li>
    <li>smoking</li>
</ul>

The categorical data uses *label encoding*

In [4]:
print("Dead Patients: ",format(len(heartFailure_df.loc[heartFailure_df['DEATH_EVENT'] == 1])))

print("Alive Patients: ",format(len(heartFailure_df.loc[heartFailure_df['DEATH_EVENT'] == 0])))

Dead Patients:  96
Alive Patients:  203


There are 96 samples that correspond to dead patients.
There are 203 samples that correspond to alive patients

In [10]:
print("Number of Men in the Dataset: ",format(len(heartFailure_df.loc[heartFailure_df['sex'] == 0])))

print("Number of Women in the Dataset: ",format(len(heartFailure_df.loc[heartFailure_df['sex'] == 1])))

Number of Men in the Dataset:  105
Number of Women in the Dataset:  194


Assuming male = 0, female = 1. There are 105 males and 194 females in this dataset.

In [11]:
print("Number of Smokers in the Dataset: ",format(len(heartFailure_df.loc[heartFailure_df['smoking'] == 1])))

print("Number of Non-Smokers in the Dataset: ",format(len(heartFailure_df.loc[heartFailure_df['smoking'] == 0])))

Number of Smokers in the Dataset:  96
Number of Non-Smokers in the Dataset:  203


Assuming 0 are non-smoker and 1 are smokers, there are 203 non smokers and 96 smokers

### 1.2

In [28]:
heartFailure_df.corr(method='pearson')

age   anaemia  creatinine_phosphokinase  \
age                       1.000000  0.088006                 -0.081584   
anaemia                   0.088006  1.000000                 -0.190741   
creatinine_phosphokinase -0.081584 -0.190741                  1.000000   
diabetes                 -0.101012 -0.012729                 -0.009639   
ejection_fraction         0.060098  0.031557                 -0.044080   
high_blood_pressure       0.093289  0.038182                 -0.070590   
platelets                -0.052354 -0.043786                  0.024463   
serum_creatinine          0.159187  0.052174                 -0.016408   
serum_sodium             -0.045966  0.041882                  0.059550   
sex                       0.065430 -0.094769                  0.079791   
smoking                   0.018668 -0.107290                  0.002421   
time                     -0.224068 -0.141414                 -0.009346   
DEATH_EVENT               0.253729  0.066270                  0.062728   

                          diabetes  ejection_fraction  high_blood_pressure  \
age                      -0.101012           0.060098             0.093289   
anaemia                  -0.012729           0.031557             0.038182   
creatinine_phosphokinase -0.009639          -0.044080            -0.070590   
diabetes                  1.000000          -0.004850            -0.012732   
ejection_fraction        -0.004850           1.000000             0.024445   
high_blood_pressure      -0.012732           0.024445             1.000000   
platelets                 0.092193           0.072177             0.049963   
serum_creatinine         -0.046975          -0.011302            -0.004935   
serum_sodium             -0.089551           0.175902             0.037109   
sex                      -0.157730          -0.148386            -0.104615   
smoking                  -0.147173          -0.067315            -0.055711   
time                      0.033726           0.041729            -0.196439   
DEATH_EVENT              -0.001943          -0.268603             0.079351   

                          platelets  serum_creatinine  serum_sodium       sex  \
age                       -0.052354          0.159187     -0.045966  0.065430   
anaemia                   -0.043786          0.052174      0.041882 -0.094769   
creatinine_phosphokinase   0.024463         -0.016408      0.059550  0.079791   
diabetes                   0.092193         -0.046975     -0.089551 -0.157730   
ejection_fraction          0.072177         -0.011302      0.175902 -0.148386   
high_blood_pressure        0.049963         -0.004935      0.037109 -0.104615   
platelets                  1.000000         -0.041198      0.062125 -0.125120   
serum_creatinine          -0.041198          1.000000     -0.189095  0.006970   
serum_sodium               0.062125         -0.189095      1.000000 -0.027566   
sex                       -0.125120          0.006970     -0.027566  1.000000   
smoking                    0.028234         -0.027414      0.004813  0.445892   
time                       0.010514         -0.149315      0.087640 -0.015608   
DEATH_EVENT               -0.049139          0.294278     -0.195204 -0.004316   

                           smoking      time  DEATH_EVENT  
age                       0.018668 -0.224068     0.253729  
anaemia                  -0.107290 -0.141414     0.066270  
creatinine_phosphokinase  0.002421 -0.009346     0.062728  
diabetes                 -0.147173  0.033726    -0.001943  
ejection_fraction        -0.067315  0.041729    -0.268603  
high_blood_pressure      -0.055711 -0.196439     0.079351  
platelets                 0.028234  0.010514    -0.049139  
serum_creatinine         -0.027414 -0.149315     0.294278  
serum_sodium              0.004813  0.087640    -0.195204  
sex                       0.445892 -0.015608    -0.004316  
smoking                   1.000000 -0.022839    -0.012623  
time                     -0.022839  1.000000  

The two variables that are the most positively correlated with *Death Event* are high_blood_pressure and anaemia. The ones that are most negatively correlated are ejection_fraction, and time.

It makes sense for anemia to be positively correlated with death from heart failure. Anaemia leads to reduced oxygen flow to the body's organs, and with a string of other issues it could very well lead to death. 

For ejection fraction it makes sense for it to have a negative correlation with *Death Event*. It's just the percentage of blood that leaves the heart after each pump, and if that changes it should be from some condition (i.e physical exercise). It is not a direct cause of heart failure. Time also makes sense for it to negatively correlate since it's just the follow-up period in days. 

### 1.3

In [41]:
heartFailure_df.describe() # general descriptive statistics of the 
                           # dataframe. Can be useful for finding outliers, or
                           # if any columns has missing values

age     anaemia  creatinine_phosphokinase    diabetes  \
count  299.000000  299.000000                299.000000  299.000000   
mean    60.833893    0.431438                581.839465    0.418060   
std     11.894809    0.496107                970.287881    0.494067   
min     40.000000    0.000000                 23.000000    0.000000   
25%     51.000000    0.000000                116.500000    0.000000   
50%     60.000000    0.000000                250.000000    0.000000   
75%     70.000000    1.000000                582.000000    1.000000   
max     95.000000    1.000000               7861.000000    1.000000   

       ejection_fraction  high_blood_pressure      platelets  \
count         299.000000           299.000000     299.000000   
mean           38.083612             0.351171  263358.029264   
std            11.834841             0.478136   97804.236869   
min            14.000000             0.000000   25100.000000   
25%            30.000000             0.000000  212500.000000   
50%            38.000000             0.000000  262000.000000   
75%            45.000000             1.000000  303500.000000   
max            80.000000             1.000000  850000.000000   

       serum_creatinine  serum_sodium         sex    smoking        time  \
count         299.00000    299.000000  299.000000  299.00000  299.000000   
mean            1.39388    136.625418    0.648829    0.32107  130.260870   
std             1.03451      4.412477    0.478136    0.46767   77.614208   
min             0.50000    113.000000    0.000000    0.00000    4.000000   
25%             0.90000    134.000000    0.000000    0.00000   73.000000   
50%             1.10000    137.000000    1.000000    0.00000  115.000000   
75%             1.40000    140.000000    1.000000    1.00000  203.000000   
max             9.40000    148.000000    1.000000    1.00000  285.000000   

       DEATH_EVENT  
count    299.00000  
mean       0.32107  
std        0.46767  
min        0.00000  
25%        0.00000  
50%        0.00000  
75%        1.00000  
max        1.00000

In [57]:
for decimal in heartFailure_df['age']:
    # some entries in the age column are not integers. The function of this 
    # loop is to check for any non-integern values in Age.
    residue = decimal - int(decimal) 
    float_residue = float(residue)
    if(float_residue != 0):
        heartFailure_df['age'].replace(decimal, np.nan, inplace = True)
        break

heartFailure_df.head(189) # testing this with rows 185 & 189 since those were
                          # rows that had non-integer values for Age.

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
1    55.0        0                      7861         0                 38   
2    65.0        0                       146         0                 20   
3    50.0        1                       111         0                 20   
4    65.0        1                       160         1                 20   
..    ...      ...                       ...       ...                ...   
184  58.0        1                       145         0                 25   
185   NaN        1                       104         1                 30   
186  50.0        0                       582         0                 50   
187  60.0        0                      1896         1                 25   
188   NaN        1                       151         1                 40   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               1.9           130    1   
1                      0  263358.03               1.1           136    1   
2                      0  162000.00               1.3           129    1   
3                      0  210000.00               1.9           137    1   
4                      0  327000.00               2.7           116    0   
..                   ...        ...               ...           ...  ...   
184                    0  219000.00               1.2           137    1   
185                    0  389000.00               1.5           136    1   
186                    0  153000.00               0.6           134    0   
187                    0  365000.00               2.1           144    0   
188                    1  201000.00               1.0           136    0   

     smoking  time  DEATH_EVENT  
0          0     4            1  
1          0     6            1  
2          1     7            1  
3          0     7            1  
4          0     8            1  
..       ...   ...          ...  
184        1   170            1  
185        0   171            1  
186        0   172            1  
187        0   172            1  
188        0   172            0  

[189 rows x 13 columns]

Anaemia column does not need cleanup as there are no values missing, and it is **categorical data**. The same applies to diabetes, high_blood_pressue, sex, and smoking since those features are also categorical

In [201]:
### Outliers identified with based on Percentiles ###

# Ignoring the categorical data. Categorical data does not have any missing values
# Age is also ignore from this as it was previously cleaned up

Q1 = heartFailure_df.quantile(0.15)
Q3= heartFailure_df.quantile(0.85)
print(Q1, '\n')
print(Q3)

age                             49.0
anaemia                          0.0
creatinine_phosphokinase        83.1
diabetes                         0.0
ejection_fraction               25.0
high_blood_pressure              0.0
platelets                   178100.0
serum_creatinine                 0.8
serum_sodium                   133.0
sex                              0.0
smoking                          0.0
time                            34.4
DEATH_EVENT                      0.0
Name: 0.15, dtype: float64 

age                             72.60
anaemia                          1.00
creatinine_phosphokinase       841.00
diabetes                         1.00
ejection_fraction               50.00
high_blood_pressure              1.00
platelets                   341000.00
serum_creatinine                 1.83
serum_sodium                   140.00
sex                              1.00
smoking                          1.00
time                           215.00
DEATH_EVENT                      1

In [58]:
# iterating through the various features that need to be cleaned up 
for value in heartFailure_df['creatinine_phosphokinase']:
    if(value < 83 or value > 841.0):
        heartFailure_df['creatinine_phosphokinase'].replace(value, np.nan, inplace = True)
    
for value in heartFailure_df['ejection_fraction']:
    if(value < 20 or value > 50.0):
        heartFailure_df['ejection_fraction'].replace(value, np.nan, inplace = True)
        
for value in heartFailure_df['platelets']:
    if(value < 178100 or value > 341000):
        heartFailure_df['platelets'].replace(value, np.nan, inplace = True)
        
for value in heartFailure_df['serum_creatinine']:
    if(value < 0.8 or value > 1.83):
        heartFailure_df['serum_creatinine'].replace(value, np.nan, inplace = True)
        
for value in heartFailure_df['serum_sodium']:
    if(value < 133 or value > 140):
        heartFailure_df['serum_sodium'].replace(value, np.nan, inplace = True)

for value in heartFailure_df['time']:
    if(value < 34 or value > 215):
        heartFailure_df['time'].replace(value, np.nan, inplace = True)
        
        
heartFailure_df.head(190)

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                     582.0         0               20.0   
1    55.0        0                       NaN         0               38.0   
2    65.0        0                     146.0         0               20.0   
3    50.0        1                     111.0         0               20.0   
4    65.0        1                     160.0         1               20.0   
..    ...      ...                       ...       ...                ...   
185   NaN        1                     104.0         1               30.0   
186  50.0        0                     582.0         0               50.0   
187  60.0        0                       NaN         1               25.0   
188   NaN        1                     151.0         1               40.0   
189  40.0        0                     244.0         0               45.0   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               NaN           NaN    1   
1                      0  263358.03               1.1         136.0    1   
2                      0        NaN               1.3           NaN    1   
3                      0  210000.00               NaN         137.0    1   
4                      0  327000.00               NaN           NaN    0   
..                   ...        ...               ...           ...  ...   
185                    0        NaN               1.5         136.0    1   
186                    0        NaN               NaN         134.0    0   
187                    0        NaN               NaN           NaN    0   
188                    1  201000.00               1.0         136.0    0   
189                    1  275000.00               0.9         140.0    0   

     smoking   time  DEATH_EVENT  
0          0    NaN            1  
1          0    NaN            1  
2          1    NaN            1  
3          0    NaN            1  
4          0    NaN            1  
..       ...    ...          ...  
185        0  171.0            1  
186        0  172.0            1  
187        0  172.0            1  
188        0  172.0            0  
189        0  174.0            0  

[190 rows x 13 columns]

In [11]:
heartFailure_df.to_csv('heart_failure_dataset_NaNs.csv', encoding='UTF-8', index=False) # create new csv file

In [59]:
fail_df = pd.read_csv('heart_failure_dataset_NaNs.csv') # reading the csv 

In [84]:
fail_df.head(190) # confirming it contains the NaN values
                  # this is the dataframe that contains the NaN values

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                     582.0         0               20.0   
1    55.0        0                       NaN         0               38.0   
2    65.0        0                     146.0         0               20.0   
3    50.0        1                     111.0         0               20.0   
4    65.0        1                     160.0         1               20.0   
..    ...      ...                       ...       ...                ...   
185   NaN        1                     104.0         1               30.0   
186  50.0        0                     582.0         0               50.0   
187  60.0        0                       NaN         1               25.0   
188   NaN        1                     151.0         1               40.0   
189  40.0        0                     244.0         0               45.0   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               NaN           NaN    1   
1                      0  263358.03               1.1         136.0    1   
2                      0        NaN               1.3           NaN    1   
3                      0  210000.00               NaN         137.0    1   
4                      0  327000.00               NaN           NaN    0   
..                   ...        ...               ...           ...  ...   
185                    0        NaN               1.5         136.0    1   
186                    0        NaN               NaN         134.0    0   
187                    0        NaN               NaN           NaN    0   
188                    1  201000.00               1.0         136.0    0   
189                    1  275000.00               0.9         140.0    0   

     smoking   time  DEATH_EVENT  
0          0    NaN            1  
1          0    NaN            1  
2          1    NaN            1  
3          0    NaN            1  
4          0    NaN            1  
..       ...    ...          ...  
185        0  171.0            1  
186        0  172.0            1  
187        0  172.0            1  
188        0  172.0            0  
189        0  174.0            0  

[190 rows x 13 columns]

For age, that feature was cleaned different than the others. The values were made invalid if they were not integers, which was the case with 2 values on rows 185 amd 188. Ages need to be integers as there is no such thing as being 60.667 years old. </br>

All of the categorical data was kept as is. As seen with *heartFailure_df.describe()* we can see that there are no missing values for them. All the values are either 1 or 0. </br>

The rest of the features were cleaned using percentiles. Originally, the values that were going to be considered invalid were anything below the 25th percentile and above the 75th percentile. This caused a considereable amount potentially valid values to be considered invalid. Thus, in order to take account those values, by looking at their distribution, using values between the 15th and 85th percentiles allowed for more potential valid values to be included, while also invalidaing any outliers that fall outside of the range. As I am not an expert in what the correct amount of any of these features should be, these are subjective and could easily change. 



### 1.4

In [84]:
# taking out the categorical features
fail2 = fail_df # create another dataframe

# normalize 
zero_mean = fail2 - fail2.mean(skipna = True)
unit_var = zero_mean / (zero_mean.std(skipna = True))

# replace the values of the numerical data without affecting the categorical data
fail2['age'] = unit_var['age']
fail2['creatinine_phosphokinase'] = unit_var['creatinine_phosphokinase']
fail2['ejection_fraction'] = unit_var['ejection_fraction']
fail2['platelets'] = unit_var['platelets']
fail2['serum_creatinine'] = unit_var['serum_creatinine']
fail2['serum_sodium'] = unit_var['serum_sodium']
fail2['time'] = unit_var['time']

fail2.head(20)

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0   1.186852        0                  1.125703         0          -1.870338   
1  -0.488903        0                       NaN         0           0.362393   
2   0.348975        0                 -0.887259         0          -1.870338   
3  -0.907842        1                 -1.048850         0          -1.870338   
4   0.348975        1                 -0.822623         1          -1.870338   
5   2.443668        1                       NaN         0           0.610474   
6   1.186852        1                 -0.425571         0                NaN   
7  -0.069964        1                 -0.107006         1                NaN   
8   0.348975        0                 -0.836474         0                NaN   
9   1.605791        1                 -0.993448         0          -0.009729   
10  1.186852        1                       NaN         0           0.362393   
11  0.097611        0                 -0.494824         0          -1.250135   
12 -1.326780        1                       NaN         0          -0.629932   
13 -0.907842        1                 -0.785688         0           0.362393   
14 -0.991629        1                       NaN         0          -0.629932   
15  1.773366        1                  0.188475         0           1.850881   
16  2.192305        1                 -0.873409         0           0.362393   
17 -1.326780        0                  1.125703         0                NaN   
18  0.767913        1                 -0.984214         0          -1.250135   
19 -1.075417        1                  1.125703         1                NaN   

    high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                     1   0.236348               NaN           NaN    1   
1                     0   0.194519         -0.164088     -0.391927    1   
2                     0        NaN          0.572890           NaN    1   
3                     0  -1.164759               NaN      0.078385    1   
4                     0   1.815777               NaN           NaN    0   
5                     1  -1.317607               NaN           NaN    1   
6                     0        NaN          0.204401      0.078385    1   
7                     0        NaN         -0.164088           NaN    1   
8                     0   0.194519          1.309868      0.548697    0   
9                     1        NaN               NaN     -1.802863    1   
10                    1        NaN               NaN           NaN    1   
11                    1  -0.069348         -0.901066      1.489321    1   
12                    0        NaN         -0.164088      0.078385    1   
13                    1   0.516569         -0.164088      0.078385    1   
14                    1        NaN         -0.532577      0.548697    0   
15                    0        NaN          0.572890     -0.391927    1   
16                    0   0.159924         -0.901066      1.489321    1   
17                    0        NaN         -1.269554           NaN    1   
18                    1  -0.476943         -0.532577      1.489321    0   
19                    0        NaN               NaN           NaN    0   

    smoking  time  DEATH_EVENT  
0         0   NaN            1  
1         0   NaN            1  
2         1   NaN            1  
3         0   NaN            1  
4         0   NaN            1  
5         1   NaN            1  
6         0   NaN            1  
7         1   NaN            1  
8         0   NaN            1  
9         1   NaN            1  
10        1   NaN            1  
11        1   NaN            1  
12        0   NaN            1  
13        0   NaN            1  
14        0   NaN            0  
15        0   NaN            1  
16        0   NaN            1  
17        0   NaN            1  
18        0   NaN            1  
19        0   NaN            1

### 1.5

In [132]:
# One Hot Encoding
fail2['DEATH_EVENT'] = dfail2['DEATH_EVENT'].astype('object')
sheer_heart_attack = 0
for i in fail2['DEATH_EVENT']:
    if (i == 0 ):
        fail2.at[sheer_heart_attack,'DEATH_EVENT'] = [1,0]
        sheer_heart_attack += 1
    elif (i == 1):
        fail2.at[sheer_heart_attack,'DEATH_EVENT'] = [0,1]
        sheer_heart_attack += 1

fail2.to_csv('heart_failure_dataset_processed.csv', encoding='UTF-8', index=False) # Export to csv

In [136]:
fail2.head(299) # One hot encoding

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    1.186852        0                  1.125703         0          -1.870338   
1   -0.488903        0                       NaN         0           0.362393   
2    0.348975        0                 -0.887259         0          -1.870338   
3   -0.907842        1                 -1.048850         0          -1.870338   
4    0.348975        1                 -0.822623         1          -1.870338   
..        ...      ...                       ...       ...                ...   
294  0.097611        0                       NaN         1           0.362393   
295 -0.488903        0                       NaN         0           0.362393   
296 -1.326780        0                       NaN         1                NaN   
297 -1.326780        0                       NaN         0           0.362393   
298 -0.907842        0                 -0.656415         0           1.230678   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1   0.236348               NaN           NaN    1   
1                      0   0.194519         -0.164088     -0.391927    1   
2                      0        NaN          0.572890           NaN    1   
3                      0  -1.164759               NaN      0.078385    1   
4                      0   1.815777               NaN           NaN    0   
..                   ...        ...               ...           ...  ...   
294                    1        NaN         -0.164088           NaN    1   
295                    0   0.363721          0.204401      1.019009    0   
296                    0        NaN         -1.269554      0.548697    0   
297                    0        NaN          0.941379      1.489321    1   
298                    0        NaN          1.678357     -0.391927    1   

     smoking  time DEATH_EVENT  
0          0   NaN      [0, 1]  
1          0   NaN      [0, 1]  
2          1   NaN      [0, 1]  
3          0   NaN      [0, 1]  
4          0   NaN      [0, 1]  
..       ...   ...         ...  
294        1   NaN      [1, 0]  
295        0   NaN      [1, 0]  
296        0   NaN      [1, 0]  
297        1   NaN      [1, 0]  
298        1   NaN      [1, 0]  

[299 rows x 13 columns]

## Problem 2

In [159]:
diabetes_df = pd.read_csv('diabetes_dataset.csv')

### 2.1

In [11]:
diabetesShape = diabetes_df.shape
diabetesAxes = diabetes_df.axes
diabetesDtype = diabetes_df.dtypes

print("The shape of the dataset: ", diabetesShape)
print("\nThe axes of the dataset: \n", diabetesAxes)
print("\nThe dtype of the dataset: \n", diabetesDtype)

The shape of the dataset:  (768, 9)

The axes of the dataset: 
 [RangeIndex(start=0, stop=768, step=1), Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')]

The dtype of the dataset: 
 Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


In [72]:
diabetes_df.head(10)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   
5            5      116             74              0        0  25.6   
6            3       78             50             32       88  31.0   
7           10      115              0              0        0  35.3   
8            2      197             70             45      543  30.5   
9            8      125             96              0        0   0.0   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  
5                     0.201   30        0  
6                     0.248   26        1  
7                     0.134   29        0  
8                     0.158   53        1  
9                     0.232   54        1

This is *supervised learning* problem. The first 8 features are independent, with the Outcome being a dependent feature on those previous ones. It is trying to predict someone having diabetes based on those various features. 

This dataset contains 8 independent variables.
<ol>
    <li>Pregnancies</li>
    <li>Glucose</li>
    <li>Blood Pressure</li>
    <li>Skin Thicknes</li>
    <li>Insulin</li>
    <li>BMI</li>
    <li>Diabetes Pedigree Function</li>
    <li>Age</li>
</ol> </br>

It only has 1 dependent variable it is *Outcome* </br>

Only 1 categorical data variable and it is *Outcome*, which uses **label encoding**

In [45]:
young = 0
middle = 0
old = 0
for i in diabetes_df['Age']:
    if(i >= 21 and i <= 40):
        young +=1 
    elif (i >= 41 and i <= 60):
        middle += 1
    elif (i >= 61):
        old += 1
    else:
        continue
        
print("Young adults (21-40): ",young)
print("Middle-aged adults (41-60): ",middle)
print("Old-aged adults (61+): ",old)

print("\nAll ages add up to: ",young + middle + old)

Young adults (21-40):  574
Middle-aged adults (41-60):  167
Old-aged adults (61+):  27

All ages add up to:  768


### 2.2

In [73]:
diabetes_df['Glucose'].replace(0, np.nan, inplace = True)
diabetes_df['BloodPressure'].replace(0, np.nan, inplace = True)
diabetes_df['SkinThickness'].replace(0, np.nan, inplace = True)
diabetes_df['Insulin'].replace(0, np.nan, inplace = True)
diabetes_df['BMI'].replace(0, np.nan, inplace = True)

# checking for any negative values
for i in diabetes_df.iloc[1,0:7]: # select "first" row (the row without column name) and columns 0-7
    if(i < 0):
        diabetes_df.replace(i, np.nan, inplace = True)
        
diabetes_df.head(20)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0             6    148.0           72.0           35.0      NaN  33.6   
1             1     85.0           66.0           29.0      NaN  26.6   
2             8    183.0           64.0            NaN      NaN  23.3   
3             1     89.0           66.0           23.0     94.0  28.1   
4             0    137.0           40.0           35.0    168.0  43.1   
5             5    116.0           74.0            NaN      NaN  25.6   
6             3     78.0           50.0           32.0     88.0  31.0   
7            10    115.0            NaN            NaN      NaN  35.3   
8             2    197.0           70.0           45.0    543.0  30.5   
9             8    125.0           96.0            NaN      NaN   NaN   
10            4    110.0           92.0            NaN      NaN  37.6   
11           10    168.0           74.0            NaN      NaN  38.0   
12           10    139.0           80.0            NaN      NaN  27.1   
13            1    189.0           60.0           23.0    846.0  30.1   
14            5    166.0           72.0           19.0    175.0  25.8   
15            7    100.0            NaN            NaN      NaN  30.0   
16            0    118.0           84.0           47.0    230.0  45.8   
17            7    107.0           74.0            NaN      NaN  29.6   
18            1    103.0           30.0           38.0     83.0  43.3   
19            1    115.0           70.0           30.0     96.0  34.6   

    DiabetesPedigreeFunction  Age  Outcome  
0                      0.627   50        1  
1                      0.351   31        0  
2                      0.672   32        1  
3                      0.167   21        0  
4                      2.288   33        1  
5                      0.201   30        0  
6                      0.248   26        1  
7                      0.134   29        0  
8                      0.158   53        1  
9                      0.232   54        1  
10                     0.191   30        0  
11                     0.537   34        1  
12                     1.441   57        0  
13                     0.398   59        1  
14                     0.587   51        1  
15                     0.484   32        1  
16                     0.551   31        1  
17                     0.254   31        1  
18                     0.183   33        0  
19                     0.529   32        1

For Glucose, Blood Pressure, Skin Thickness, Insuling and BMI, I concluded that any valid value in those features could not be 0. Therefore, they were replaced with *NaN*. Pregnancies can have valid values of 0 so those were left alone. Due to being unfamiliar with Diabetes Pedigree Function, it was also not checked for 0 values since those could possibly be valid. </br>

Age did not need to be touched since there are no values of Age less than 21, this was checked in the previous cells when accounting for Age ranges and they all add up to 768. </br>

Furthermore, since I cannot see every value, I also decided to check for any non-positive value that could be sneaking around in the dataset. Non-positive values do not have any place in this dataset 

### 2.3

In [160]:
# Normalizing the dataset
diabetes_mean = diabetes_df.loc[:,'Pregnancies':'Age'] - diabetes_df.loc[:,'Pregnancies':'Age'].mean(skipna = True)
diabetes_var = diabetes_mean / (diabetes_mean.std(skipna = True))

diabetes_df['Pregnancies'] = diabetes_var['Pregnancies'] 
diabetes_df['Glucose'] = diabetes_var['Glucose'] 
diabetes_df['BloodPressure'] = diabetes_var['BloodPressure'] 
diabetes_df['SkinThickness'] = diabetes_var['SkinThickness'] 
diabetes_df['Insulin'] = diabetes_var['Insulin'] 
diabetes_df['BMI'] = diabetes_var['BMI'] 
diabetes_df['DiabetesPedigreeFunction'] = diabetes_var['DiabetesPedigreeFunction'] 
diabetes_df['Age'] = diabetes_var['Age'] 

diabetes_df.head(20)

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0      0.639530  0.847771       0.149543       0.906679 -0.692439  0.203880   
1     -0.844335 -1.122665      -0.160441       0.530556 -0.692439 -0.683976   
2      1.233077  1.942458      -0.263769      -1.287373 -0.692439 -1.102537   
3     -0.844335 -0.997558      -0.160441       0.154433  0.123221 -0.493721   
4     -1.141108  0.503727      -1.503707       0.906679  0.765337  1.408828   
5      0.342757 -0.153085       0.252871      -1.287373 -0.692439 -0.810813   
6     -0.250789 -1.341602      -0.987066       0.718617  0.071158 -0.125895   
7      1.826623 -0.184362      -3.570271      -1.287373 -0.692439  0.419502   
8     -0.547562  2.380333       0.046215       1.533551  4.019303 -0.189314   
9      1.233077  0.128406       1.389481      -1.287373 -0.692439 -4.057829   
10     0.045984 -0.340746       1.182825      -1.287373 -0.692439  0.711226   
11     1.826623  1.473307       0.252871      -1.287373 -0.692439  0.761961   
12     1.826623  0.566280       0.562856      -1.287373 -0.692439 -0.620558   
13    -0.844335  2.130119      -0.470426       0.154433  6.648507 -0.240048   
14     0.342757  1.410753       0.149543      -0.096316  0.826078 -0.785445   
15     0.936304 -0.653513      -3.570271      -1.287373 -0.692439 -0.252732   
16    -1.141108 -0.090532       0.769512       1.658926  1.303326  1.751286   
17     0.936304 -0.434576       0.252871      -1.287373 -0.692439 -0.303466   
18    -0.844335 -0.559683      -2.020348       1.094741  0.027772  1.434195   
19    -0.844335 -0.184362       0.046215       0.593243  0.140576  0.330716   

    DiabetesPedigreeFunction       Age  Outcome  
0                   0.468187  1.425067        1  
1                  -0.364823 -0.190548        0  
2                   0.604004 -0.105515        1  
3                  -0.920163 -1.040871        0  
4                   5.481337 -0.020483        1  
5                  -0.817546 -0.275580        0  
6                  -0.675693 -0.615709        1  
7                  -1.019762 -0.360612        0  
8                  -0.947326  1.680164        1  
9                  -0.723983  1.765196        1  
10                 -0.847727 -0.275580        0  
11                  0.196553  0.064549        1  
12                  2.924962  2.020293        0  
13                 -0.222970  2.190358        1  
14                  0.347461  1.510099        1  
15                  0.036591 -0.105515        1  
16                  0.238807 -0.190548        1  
17                 -0.657584 -0.190548        1  
18                 -0.871873 -0.020483        0  
19                  0.172408 -0.105515        1

### 2.4

In [157]:
# One Hot Encoded dataframe
diabetes_df2 = diabetes_df
diabetes_df2['Outcome'] = diabetes_df2['Outcome'].astype('object')
killer_queen = 0
for i in diabetes_df2['Outcome']:
    if (i == 0 ):
        diabetes_df2.at[killer_queen,'Outcome'] = [1,0]
        killer_queen += 1
    elif (i == 1):
        diabetes_df2.at[killer_queen,'Outcome'] = [0,1]
        killer_queen += 1

diabetes_df.head(20)



Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0      0.639530  0.861722      -0.032723       0.558040       NaN  0.164987   
1     -0.844335 -1.201441      -0.517291      -0.014643       NaN -0.845845   
2      1.233077  2.007924      -0.678814            NaN       NaN -1.322380   
3     -0.844335 -1.070446      -0.517291      -0.587327 -0.518188 -0.629238   
4     -1.141108  0.501487      -2.617087       0.558040  0.104834  1.536831   
5      0.342757 -0.186234       0.128800            NaN       NaN -0.990249   
6     -0.250789 -1.430681      -1.809473       0.271699 -0.568703 -0.210464   
7      1.826623 -0.218982            NaN            NaN       NaN  0.410475   
8     -0.547562  2.466404      -0.194246       1.512514  3.262042 -0.282667   
9      1.233077  0.108504       1.905550            NaN       NaN       NaN   
10     0.045984 -0.382725       1.582504            NaN       NaN  0.742606   
11     1.826623  1.516694       0.128800            NaN       NaN  0.800368   
12     1.826623  0.566985       0.613368            NaN       NaN -0.773642   
13    -0.844335  2.204415      -1.001860      -0.587327  5.813065 -0.340429   
14     0.342757  1.451197      -0.032723      -0.969117  0.163769 -0.961368   
15     0.936304 -0.710211            NaN            NaN       NaN -0.354869   
16    -1.141108 -0.120736       0.936413       1.703408  0.626826  1.926723   
17     0.936304 -0.480971       0.128800            NaN       NaN -0.412631   
18    -0.844335 -0.611966      -3.424701       0.844382 -0.610799  1.565712   
19    -0.844335 -0.218982      -0.194246       0.080804 -0.501350  0.309392   

    DiabetesPedigreeFunction       Age Outcome  
0                   0.468187  1.425067  [1, 0]  
1                  -0.364823 -0.190548  [0, 1]  
2                   0.604004 -0.105515  [1, 0]  
3                  -0.920163 -1.040871  [0, 1]  
4                   5.481337 -0.020483  [1, 0]  
5                  -0.817546 -0.275580  [0, 1]  
6                  -0.675693 -0.615709  [1, 0]  
7                  -1.019762 -0.360612  [0, 1]  
8                  -0.947326  1.680164  [0, 1]  
9                  -0.723983  1.765196  [1, 0]  
10                 -0.847727 -0.275580  [0, 1]  
11                  0.196553  0.064549  [1, 0]  
12                  2.924962  2.020293  [0, 1]  
13                 -0.222970  2.190358  [0, 1]  
14                  0.347461  1.510099  [0, 1]  
15                  0.036591 -0.105515  [0, 1]  
16                  0.238807 -0.190548  [0, 1]  
17                 -0.657584 -0.190548  [1, 0]  
18                 -0.871873 -0.020483  [0, 1]  
19                  0.172408 -0.105515  [1, 0]

### 2.5

In [162]:
# For outcome with values 0
zero_outcome = diabetes_df.loc[diabetes_df['Outcome'] == 0]
zero_median = zero_outcome.median(skipna = True)
zero_replace = zero_outcome.fillna(zero_median)

# For outcome with value 1
one_outcome = diabetes_df.loc[diabetes_df['Outcome'] == 1]
one_median = one_outcome.median(skipna = True)
one_replace = one_outcome.fillna(one_median)

# Combining the two separate datasets
diabetes_combine = [zero_replace, one_replace]
diabetes_median = pd.concat(diabetes_combine)
final_diabetes_dataset = diabetes_median.sort_index(axis = 0)
final_diabetes_dataset

Pregnancies   Glucose  BloodPressure  SkinThickness   Insulin       BMI  \
0       0.639530  0.847771       0.149543       0.906679 -0.692439  0.203880   
1      -0.844335 -1.122665      -0.160441       0.530556 -0.692439 -0.683976   
2       1.233077  1.942458      -0.263769      -1.287373 -0.692439 -1.102537   
3      -0.844335 -0.997558      -0.160441       0.154433  0.123221 -0.493721   
4      -1.141108  0.503727      -1.503707       0.906679  0.765337  1.408828   
..           ...       ...            ...            ...       ...       ...   
763     1.826623 -0.622237       0.356200       1.721613  0.869464  0.115094   
764    -0.547562  0.034575       0.046215       0.405181 -0.692439  0.609757   
765     0.342757  0.003299       0.149543       0.154433  0.279412 -0.734711   
766    -0.844335  0.159683      -0.470426      -1.287373 -0.692439 -0.240048   
767    -0.844335 -0.872451       0.046215       0.655930 -0.692439 -0.201997   

     DiabetesPedigreeFunction       Age  Outcome  
0                    0.468187  1.425067        1  
1                   -0.364823 -0.190548        0  
2                    0.604004 -0.105515        1  
3                   -0.920163 -1.040871        0  
4                    5.481337 -0.020483        1  
..                        ...       ...      ...  
763                 -0.908090  2.530487        0  
764                 -0.398023 -0.530677        0  
765                 -0.684747 -0.275580        0  
766                 -0.370859  1.169970        1  
767                 -0.473476 -0.870806        0  

[768 rows x 9 columns]

In [164]:
# One hot encoding the last dataset
final_diabetes_dataset['Outcome'] = final_diabetes_dataset['Outcome'].astype('object')
killer_queen = 0
for i in final_diabetes_dataset['Outcome']:
    if (i == 0 ):
        final_diabetes_dataset.at[killer_queen,'Outcome'] = [1,0]
        killer_queen += 1
    elif (i == 1):
        final_diabetes_dataset.at[killer_queen,'Outcome'] = [0,1]
        killer_queen += 1

final_diabetes_dataset.to_csv('diabetes_dataset_processed.csv', encoding='UTF-8', index=False) # Export to csv